---
title: "Peak Calling & Peak2Gene Linkage"
output: 
  html_document:
    toc: true
    toc_depth: 5
    code_folding: hide
    toc_float: true
    code_download: true
    theme: cosmo
    highlight: textmate
---

<style>
body {
text-align: justify}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = FALSE, autodep = TRUE, 
                      collapse = TRUE, message = FALSE)
knitr::opts_knit$set(root.dir = "/omics/groups/OE0533/internal/katharina/scDoRI/gastrulation_data/")
setwd("/omics/groups/OE0533/internal/katharina/scDoRI/gastrulation_data/")

set.seed(1)
```

  
```{r}
## Load libraries


suppressPackageStartupMessages({library(ArchR)
library(knitr)
library(rhdf5)
library(uwot)
library(tidyverse)
library(scater)
library(reticulate)
library(dichromat)
library(zellkonverter)
#library(caret)
h5disableFileLocking()
})
```


```{r}
# load ArchR project
proj <- loadArchRProject(path = "10_peaks_called/")

# for R to find macs2 we will use a conda environment where the latest version 
# of macs2 is installed
conda_list()[[1]][2] %>% use_condaenv(required = TRUE)
```



# Pseudo-bulk replicates

For statistical significance we need replicates, which is not given 
in the binary single cell data obtained from scATAC-seq. Therefore,
we create pseudo-bulk replicates. We group single cells and the data 
of all the cells in the group are combined into a pseudo-sample that 
resembles a bulk-ATAC-seq experiment. For each cell grouping several 
pseudo-bulks are created, resulting in pseudo-bulk replicates. The main
assumption here is that the cells we group together are so similar that
we are not interested in the differences between them. We group cells
together based on clusters and known cell identities. 

Whenever possible ArchR attempts to create pseudo-bulk replicates in a sample-aware
fashion, where ch replicate only contains cells from a single sample. If this is not
possible the requirements become successively less stringent.

Parameters:

* minReps, maxReps: min. and max. number of replicates
* minCells, maxCells: min. and max. number of cells per replicate
* SamplingRatio: sampling ratio to use if a particular group of cells
does not have required number
e.g. 0.8 means that 80% of cells can be sampled without replacement for a particular replicate

Using the generated pseudo-bulk replicates, peak calling can be done. 

The `groupBy` parameter specifies for which groups pseudo-bulk replicates should be made. With pseudo-bulks you can call peaks next, 
because we do not want to call peaks on all single cells together. 

How many cells do we have for each celltype?

```{r, results = "asis"}
getCellColData(proj) %>% as.data.frame() %>% 
  mutate(timepoint = str_remove(Sample, "_rep[1-2]")) %>% 
  group_by(celltypes, timepoint) %>% summarize(n = n()) %>% knitr::kable()
```


```{r, results = "asis"}
getCellColData(proj) %>% as.data.frame() %>% 
  mutate(timepoint = str_remove(Sample, "_rep[1-2]")) %>% 
  group_by(celltypes) %>% summarize(n = n()) %>% knitr::kable()
```




```#{r}
proj <- addGroupCoverages(
  ArchRProj = proj,
  groupBy = "celltypes",
  useLabels = TRUE,
  minCells = 40,
  maxCells = 500,
  maxFragments = 25 * 10^6,
  minReplicates = 2,
  maxReplicates = 5,
  sampleRatio = 0.8,
  kmerLength = 6,
  threads = getArchRThreads(),
  returnGroups = FALSE,
  parallelParam = NULL,
  force = FALSE,
  verbose = TRUE,
  logFile = createLogFile("addGroupCoverages")
)
```

```{r}
#saveArchRProject(proj, "09_group_coverage")
```



# Peak calling
https://hbctraining.github.io/Intro-to-ChIPseq/lessons/05_peak_calling_macs.html

Because the scATAC-seq data is binary, peaks cannot be called on single 
cells. Using pseuodo-bulk replicates we can assess the reproducibility 
of our peak calls. 

ArchR uses a fixed peak width as opposed to a variable peak width, 
because then peak length does not need to be normalized and it is
easier to merge peaks from different saples. The peak length used is 
501bp, since most peaks are smaller than 501 bp. 


## Iterative Overlap

1. rank peaks by their signficance
2. the most significant peak is retained
3. any peak overlapping with the most significant peak is removed
4. the process is repeated for the remaining peaks


For example, if we hae 3 ell types A, B, C and 3 pseudo-bulk replicates for each:

1. Call peaks for each pseudo-bulk replicate individually
2. The first iteration of iterative overlap would then be performed on all pseuo-
bulk replicates from a single cell type. (It is important to normalize peak 
significance across different samples beforehand)
3. Check the reproducibility of each peak across pseud9-bulk replicates and only
keep the peaks that pass a threshold (this threshold is indicated by the `reproducibility`
parameter).
4. You now have a single merged peak set for each of the 3 cell types
5. proceed with iterative overlap remove to merge A, B, C peak sets (It is 
important to re-normalize the peak significance across different cell types at 
this step)

**At the end you are left with a single merged peak set of fixed-width peaks.**

Cell are then defined as accessible at a particular peak if a read from that 
cell overlapped the peak. 

We set the minimum number of cells used to call a peak to 25, this way
we will exclude pseudo-bulk replicates with very low cell number.

Here I set the reproducibility parameter to 2. Any peaks below that will be 
removed. This means that at least two samples must have a peak call at this locus.
A maximum of 500 peaks can be called per cell. A maximum of 150,000 peaks can be called
per group. The cells are grouped by celltype prior to calling peaks. 

The q-value cutoff for signficance testing during peak calling is 0.1. The q-value 
gives the expected positive FDR obtained by rejecting the null hypothesis (proportion
of false discoveries among all positive results. If we set a cutoff at 0.1, the 
expected false discovery rate is 0.1.

In our case peaks within 2000bp upstream and 100bp downstream of the TSS 
are annotated as a promoter peak. 

If you use `getGeneAnnotations()` you can get a list of gene annotations (including
coordinates, gene-id and symbols/names) for exons, TSS and genes. 

```{r}
addArchRThreads(1)
#pathToMacs2 <- findMacs2()
```


```#{r}
proj <- addReproduciblePeakSet(
    ArchRProj = proj, 
    groupBy = "celltypes",
    pathToMacs2 = "/omics/groups/OE0533/internal/katharina/R_env/bin/macs2",
    minCells = 25, # minimum allowable number of unique cells that was used to 
    #create the coverage files on which peaks are called. Pseudo-bulk replicates 
    #derived from very low cell numbers are excluded.
    force = FALSE
)
```



## Peak Statistics

we should probably try to filter peaks which get a low score and genes 
with low variability. 


```{r}
peakset <- ArchR::getPeakSet(proj) %>%
  tibble::as_tibble()
print(paste0("Dimensions of the peak set are: ", paste(dim(peakset), collapse=", ")))
```



```{r}
peakset %>%
  glimpse()
```


```{r}
peakset %>%
  dplyr::count(peakType)
```

Not sure what these plots mean?

```{r}
peakset %>%
  ggplot() +
  geom_histogram(aes(x=groupScoreQuantile), bins=100, fill = "blue", 
                 color = "black", alpha = 0.6)
```

In our data the min score value is 1, while in Ricards data the min value is 3. We 
could remove an cells with values lower than 3.

```{r}
peakset %>%
  ggplot() +
  geom_histogram(aes(x=score), bins=100, fill = "blue", 
                 color = "black", alpha = 0.6)


# peakset %>%
#   ggplot() +
#   geom_histogram(aes(x=score), bins=100, fill = "blue", 
#                  color = "black", alpha = 0.6) +
#   scale_y_log10()
# 
# peakset %>%
#   ggplot() +
#   geom_histogram(aes(x=score), bins=100, fill = "blue", 
#                  color = "black", alpha = 0.6) +
#   scale_x_log10()

min(peakset$score)
```


The cutoff for the reproducibility was set to 2, meaning that if a peak 
was not found in more than two pseudo-bulk replicates it was removed.

```{r}

peakset %>%
  ggplot() +
  geom_histogram(aes(x=Reproducibility), bins=4, fill = "blue", 
                 color = "black", alpha = 0.6)
```



Lets have a look for which time stages there where enough cells to call peaks and 
how many peaks were called in each case. 

```{r}
peakset %>% group_by(GroupReplicate) %>% summarise(n = n())
```


TODO: normalize this plot

```{r, fig.width=15}
peakset <- as_data_frame(peakset)

p1 <- peakset %>% 
  mutate(Group = str_extract(peakset[["GroupReplicate"]], "(?<!.)[^.]+")) %>% 
  ggplot() +
  geom_bar(aes(x = Group, fill = peakType), position = "dodge") +
  labs(title = "type of peak") +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.8, hjust=1)) 
p1
```


```{r, fig.height=8, fig.width=8}
p2 <- peakset %>% 
  ggplot() +
  geom_violin(aes(x = peakType, y = distToGeneStart, fill = peakType),
              alpha = .5) +
  geom_boxplot(aes(x = peakType, y = distToGeneStart, fill = peakType),
               alpha = .1) +
  labs(title = "distance to gene start")

p3 <- peakset %>% 
  ggplot() +
  geom_violin(aes(x = peakType, y = distToGeneStart, fill = peakType), outlier.shape = NA,
              alpha = .5) +
  geom_boxplot(aes(x = peakType, y = distToGeneStart, fill = peakType), outlier.shape = NA,
               alpha = .1) +
  ylim(0, 500) +
  labs(title = "distance to gene start (outliers removed)")

p4 <- peakset %>% 
  ggplot() +
  geom_violin(aes(x = peakType, y = distToTSS, fill = peakType),
                  alpha = .5) + 
  geom_boxplot(aes(x = peakType, y = distToTSS, fill = peakType), alpha = .1) + 
  labs(title = "distance to TSS")

p5 <- peakset %>% 
  ggplot() +
  geom_violin(aes(x = peakType, y = distToTSS, fill = peakType),
                  alpha = .5, outlier.shape = NA ) +
  geom_boxplot(aes(x = peakType, y = distToTSS, fill = peakType), 
               alpha = .1, outlier.shape = NA) +
  ylim(0, 1500) +
  labs(title = "distance to TSS (outliers removed")

p6 <- peakset %>% 
  mutate(Group = str_extract(peakset[["GroupReplicate"]], "(?<!.)[^.]+")) %>% 
  ggplot() +
  geom_histogram(aes(x = GC), bins = 100) +
  labs(title = "GC content")

p7 <- peakset %>% 
  mutate(Group = str_extract(peakset[["GroupReplicate"]], "(?<!.)[^.]+")) %>% 
  ggplot() +
  geom_violin(aes(x = Group, y = GC, fill = Group), alpha = .5) +
  geom_boxplot(aes(x = Group, y = GC, fill = Group), alpha = .1) +
  labs(title = "GC content")


gridExtra::grid.arrange(p2, p3, p4, p5, p6, ncol = 2)
```



```{r, fig.height=8, fig.width=15}
p7

```




## Add the peak matrix

 We will add the peak matrix to the ArchR object.
 
```#{r}
proj <- addPeakMatrix(ArchRProj = proj)
```
 
```{r}
#saveArchRProject(proj, outputDirectory = "04_peak_calling_ArchRobject", load = FALSE)
```

What are our available matrices after this step?

```{r}
getAvailableMatrices(proj)
```


# Filter the peak matrix


```{r}
peakMatrix <- getMatrixFromProject(ArchRProj = proj,
                                   useMatrix = "PeakMatrix",
                                   verbose = FALSE)


print(paste0("Dimensions of the peak matrix are:", paste(dim(peakMatrix), collapse = ", ")))
```


## Overlap of peaks 

In the processed datset we got, there are no peak scores below 3 included, which 
is why we will also remove those.

```{r}
atacSE <- readH5AD("jupyter_notebooks/anndata_atac_peak_matrix.h5ad")
atacSE <- atacSE[, proj$cellNames]
```

```{r}
p1 <- rowData(atacSE) %>% as.data.frame() %>%
  ggplot() +
  geom_histogram(aes(x=score), bins=100, fill = "blue", 
                 color = "black", alpha = 0.6) +
  labs(title = "Ricards peak scores")

p2 <- peakset %>%
  ggplot() +
  geom_histogram(aes(x=score), bins=100, fill = "blue", 
                 color = "black", alpha = 0.6) +
  labs(title = "ArchR peak score")

gridExtra::grid.arrange(p1, p2, ncol = 2)
```

```{r}
# combining chromsome and idx information in one column for our ArchrProject
peakset <- peakset %>% as.data.frame() %>%
  tidyr::unite(merged_peaks, start, end, sep = "_", remove = FALSE) %>% 
  tidyr::unite(chr_idx, seqnames, idx, sep = "_", remove = FALSE)

# do the same for the processed data
processed_peaks <- as.data.frame(rowData(atacSE))

processed_peaks <- processed_peaks %>%
  tidyr::unite(merged_peaks, start, end, sep = "_", remove = FALSE) %>%
    tidyr::unite(chr_idx, chr, idx, sep = "_", remove = FALSE)

```   





## Filter peak score

First we will remove any peaks with scores below 3 and above 300. 

```{r}
peakset_filt <- peakset %>% as.data.frame() %>% filter(score > 3 & score < 300)

print(paste0("Before filtering for scores below 3 and smaller than 300 there are ", dim(peakset)[[1]], " peaks."))
print(paste0("After filtering to keep only peaks with score bigger than 3, ", dim(peakset_filt)[[1]], " peaks left."))
```

```{r}
p1 <- rowData(atacSE) %>% as.data.frame() %>%
  ggplot() +
  geom_histogram(aes(x=score), bins=100, fill = "blue", 
                 color = "black", alpha = 0.6) +
  labs(title = "Ricards peak scores")

p2 <- peakset_filt %>%
  ggplot() +
  geom_histogram(aes(x=score), bins=100, fill = "blue", 
                 color = "black", alpha = 0.6) +
  labs(title = "ArchR peak score")

gridExtra::grid.arrange(p1, p2, ncol = 2)
```






## Marker Peaks

To find variable features we can use the `getMArkerFeatures` function.  
To account for differences in data quality the bias parameter can be set
to account for TSS enrichment score and number of fragments. 

```{r}
markerPeaks_celltpyes <- getMarkerFeatures(
  ArchRProj = proj,
  groupBy = "celltypes",
  useMatrix = "PeakMatrix", 
  bias = c("TSSEnrichment", "log10(nFrags)"),
  testMethod = "wilcoxon",
  maxCells = 500,
  scaleTo = 1e4, # each column of the peak matrix will be scaled to a column sum of 1e4
  k = 100, # the number of nearby cells to use for selecting a background of bias-matched cells
  bufferRatio = .8, # select the top 80 best matched cells out of the 100 nearest neighbors
  binarize = TRUE # indicates whether a peak is present in a cell or not
)


#writeH5AD(markerPeaks_celltpyes, "marker_peaks_celltypes", assays = TRUE)

#remove anything with FDR under .01 and Log2FC below 1:
#markerPeaks_celltypes_subset <- markerPeaks_celltpyes[]

#print(paste0("The number of marker peaks when finding peaks for individual celltypes is: ", ))
```



```{r}
markerPeaks_celltypes_list <- getMarkers(markerPeaks_celltpyes,
                                         cutOff = "FDR <= .01 & Log2FC >= 2 & Pval <= .01")
markerPeaks_celltypes_list


print <- map(seq.int(1, 37), function(n){
  print(paste0(names(markerPeaks_celltypes_list)[[n]], " has ", 
                     dim(markerPeaks_celltypes_list[[n]])[[1]], " marker peaks."))
  })

# combine chromsome and index information
peak_index <- map(seq.int(1, 37), function(n){
  marker_peaks <- as.data.frame(markerPeaks_celltypes_list[[n]]) %>% 
    tidyr::unite(chr_index, seqnames, idx, sep = "_")
  peaks <- marker_peaks$chr_index
})

peak_index_list <- unlist(peak_index)
```

```{r}
print(paste0("If we filter for peaks with FDR <= 0.01, log2FC >= 2 and p-value <= 0.01 we are left with ", length(unique(peak_index_list)), " peaks"))
```


## Filtering marker peaks

What is the number of unique peaks in our peakset after filtering for scores?

```{r}
print(paste0("We have ", length(unique(as.data.frame(peakset_filt)$merged_peaks)), 
" unique peaks after filtering for peaks with scores > 3"))
```


We can now filter our peakset to contain only peaks which were found to be 
marker peaks with FDR <= 0.01, log2FC >= 2 and p-value <= 0.01. 

```{r}
peakset_final <- peakset_filt %>% as.data.frame() %>% filter(chr_idx %in% unique(peak_index_list))

print(paste0("We are now left with a peakset of ", dim(peakset_final)[[1]], " peaks."))
```
```{r}
p1 <- rowData(atacSE) %>% as.data.frame() %>%
  ggplot() +
  geom_histogram(aes(x=score), bins=100, fill = "blue", 
                 color = "black", alpha = 0.6) +
  labs(title = "Ricards peak scores")

p2 <- peakset_final %>%
  ggplot() +
  geom_histogram(aes(x=score), bins=100, fill = "blue", 
                 color = "black", alpha = 0.6) +
  labs(title = "ArchR peak score after filtering")

gridExtra::grid.arrange(p1, p2, ncol = 2)
```

### Filter Peak Matrix


```#{r}
dim(peakset_final)

peakMatrix


peak_mat <- assays(peakMatrix)[[1]]
rownames(peak_mat) <- rowData(peakMatrix)$idx

filt_peak_mat <- peak_mat[peakset_final$idx, ]
```

```{r}
#saveArchRProject(proj, "10_1_added_marker_peaks", load=FALSE)
```


## Visualize Marker Peaks

The heatmap below shows the marker peaks found for each celltype with a FDR <= 0.01 
and a Log2FC > 2.

```{r, fig.width=15, fig.height=15}
heatmapPeaks <- plotMarkerHeatmap(
  seMarker = markerPeaks_celltpyes,
  cutOff = "FDR <= .01 & Log2FC >= 2", 
  transpose = FALSE
)

draw(heatmapPeaks, heatmap_legend_side = "bot", annotation_legend_side = "bot")

```



## Find Marker Peaks for time points and celltypes simultaneously

```{r}
new_coldata <- getCellColData(proj) %>% as.data.frame() %>% mutate(timepoint = str_remove(Sample, "_rep[1-2]")) %>% 
  tidyr::unite("time_celltype", timepoint, celltypes, sep = "_", remove = FALSE) 


proj <- addCellColData(
  ArchRProj = proj,
  data = setNames(new_coldata$time_celltype, as.character(rownames(new_coldata)) ),
  name = "time_celltype", 
  cells = rownames(new_coldata), 
  force = FALSE
)

```

### Blood cells

Create a subset project which contains only blood cells:

```{r, results = "asis"}
# get the celltypes corresponding to blood
blood_cells <- getCellColData(proj) %>% as.data.frame() %>% rownames_to_column("cell") %>% 
  filter(celltypes %in% c("Erythroid1", "Erythroid2", "Erythroid3", 
                          "Blood_progenitors_1", "Blood_progenitors_2"))

# subset ArchR object to contain only blood cells
subset_archr <- subsetArchRProject(
  ArchRProj = proj,
  cells = blood_cells$cell,
  outputDirectory = "Archr_blood_cells"
)

# Have a look at the number of cells for each celltype and timepoint
blood_cells %>% group_by(time_celltype) %>% summarise(n = n()) %>% knitr::kable()
```

Compute marker peaks between one blood group and all others:

```{r}
markerPeaks_timepoint <- getMarkerFeatures(  ArchRProj = subset_archr,
  groupBy = "time_celltype",
  useMatrix = "PeakMatrix", 
  bias = c("TSSEnrichment", "log10(nFrags)"),
  testMethod = "wilcoxon",
  maxCells = 500,
  scaleTo = 1e4, # each column of the peak matrix will be scaled to a column sum of 1e4
  k = 100, # the number of nearby cells to use for selecting a background of bias-matched cells
  bufferRatio = .8, # select the top 80 best matched cells out of the 100 nearest neighbors
  binarize = TRUE # indicates whether a peak is present in a cell or not)
)
```


Below you can see a heatmap of the marker peaks in different blood cell groups 
at different time points:

```{r}
heatmapPeaks <- plotMarkerHeatmap(
  seMarker = markerPeaks_timepoint,
  cutOff = "FDR <= .01 & Log2FC >= 1", 
  transpose = FALSE, 
  nLabel = 15
)

draw(heatmapPeaks, heatmap_legend_side = "bot", annotation_legend_side = "bot")
```


TODO: It might be useful to compute pairwise comparisons between peaks. 


### Peaks for separate timepoint across all cells

```{r}
# get the different time-celltype combinations with a number of cells above 30
timepoints <- (getCellColData(proj) %>% as.data.frame() %>%  group_by(time_celltype) %>% summarise(n = n()) %>% 
  filter(n > 30)) [["time_celltype"]]

# get the cells corresponding to these timepoints with at least 30 cells
cells <- rownames(getCellColData(proj) %>% as.data.frame() %>% 
                    filter(time_celltype %in% timepoints))
```


```{r}
# all time points
markerPeaks_all_timepoint <- getMarkerFeatures(  ArchRProj = proj[cells,], # subset based on cells obtained above
  groupBy = "time_celltype",
  useMatrix = "PeakMatrix", 
  bias = c("TSSEnrichment", "log10(nFrags)"),
  testMethod = "wilcoxon",
  maxCells = 500,
  scaleTo = 1e4, # each column of the peak matrix will be scaled to a column sum of 1e4
  k = 100, # the number of nearby cells to use for selecting a background of bias-matched cells
  bufferRatio = .8, # select the top 80 best matched cells out of the 100 nearest neighbors
  binarize = TRUE # indicates whether a peak is present in a cell or not)
)
```

Below you can see all marker peaks for different time points and different 
celltypes for peaks with FDR <= 0.01 and Log2FC >= 2.

```{r, fig.width=15, fig.height=10}
heatmapPeaks <- plotMarkerHeatmap(
  seMarker = markerPeaks_all_timepoint,
  cutOff = "FDR <= .01 & Log2FC >= 2", 
  transpose = FALSE, 
  nLabel = 15
)

draw(heatmapPeaks, heatmap_legend_side = "bot", annotation_legend_side = "bot")
```








# Peak to Gene Linkage

Getting peaks2gene links took less than 5 minutes.

Manu uses all peaks on the same chromosome. The question is whether we would
like to remove peaks prior to this step based on variability between celltypes 
for example. 

When adding the peak-to-gene links, cells are first aggregated into groups. For
this a KNN graph is built. Then the 100 nearest neighbors are used to create single 
cell groups. These cell groups are therefore based on similarity between cells. The
maximum distance allowed between a peak and a gene is 250,000 by default, however
here we try to not set a limit, because Manu also considers all peaks which are on
the same genome. 

Add the filtered peakset to the object:

```{r}
proj_filt <- loadArchRProject(path = "10_peaks_called/")
```


```{r}
# get granges object from the filtered peakset
peaks_granges <- DataFrame(peakset_final %>%
                  select(seqnames, start, end, strand, score)) %>% GRanges()




proj_filt <- addPeakSet(
  ArchRProj = proj_filt,
  peakSet = peaks_granges,
  genomeAnnotation = getGenomeAnnotation(proj_filt),
  force = TRUE
)

```

Add the peak matrix:

```{r}
proj_filt <- addPeakMatrix(ArchRProj = proj_filt)
```

Significant links which are kept out of all potential peak-to-gene links
have R > 0.45 and FDR < 0.1. 

```{r}

proj_filt <- addPeak2GeneLinks(  ArchRProj = proj_filt,
  reducedDims  = "atac_LSI_100000",
  useMatrix = "GeneExpressionMatrix",
  maxDist = 250000,
  verbose = FALSE, 
  )
```


```{r}
#filt_proj <- loadArchRProject("10_2_peak2gene_links_filtered_peaks")
```


Lets have a look at the peak2gene links:

```{r}
p2g <- getPeak2GeneLinks(
    ArchRProj = filt_proj,
    corCutOff = 0.45,
    resolution = 1,
    returnLoops = FALSE
)
```



```{r, results = "asis"}
# convert the output to a tibble
p2g %>% as_tibble() %>% head()# %>% knitr::kable()
```

```{r, fig.width=10}
p1 <- p2g %>% as_tibble() %>% ggplot() +
  geom_histogram(aes(x = Correlation), bins = 100, fill = "blue",
                 color = "black", alpha = .6) 

p2 <- p2g %>% as_tibble() %>% ggplot() +
  geom_histogram(aes(x = Correlation), bins = 100, fill = "blue",
                 color = "black", alpha = .6) +
  scale_y_log10() +
  labs(y = "log10(counts)")

gridExtra::grid.arrange(p1, p2, ncol = 2)
```




```{r}
peakset %>% group_by(idx) %>% mutate(count = n()) %>% 
  arrange(count)
```


```{r}
peakset %>% group_by(merged_peaks) %>% mutate(count = n()) %>% arrange(desc(count))
```




```{r}
colPalette_celltypes = c('#532C8A',
 '#c19f70',
 '#f9decf',
 '#c9a997',
 '#B51D8D',
 '#3F84AA',
 '#9e6762',
 '#354E23',
 '#F397C0',
 '#ff891c',
 '#635547',
 '#C72228',
 '#f79083',
 '#EF4E22',
 '#989898',
 '#7F6874',
 '#8870ad',
 '#647a4f',
 '#EF5A9D',
 '#FBBE92',
 '#139992',
 '#cc7818',
 '#DFCDE4',
 '#8EC792',
 '#C594BF',
 '#C3C388',
 '#0F4A9C',
 '#FACB12',
 '#8DB5CE',
 '#1A1A1A',
 '#C9EBFB',
 '#DABE99',
 '#65A83E',
 '#005579',
 '#CDE088',
 '#f7f79e',
 '#F6BFCB')

celltypes <- (as.data.frame(getCellColData(filt_proj)) %>% group_by(celltypes) %>% 
 summarise(n = n()))$celltypes

col <- setNames(colPalette_celltypes, celltypes)
```


Eeach celltype has a different number of cells

```{r}
getCellColData(proj) %>% as.data.frame()%>% ggplot() +
  geom_bar(aes(x = celltypes, fill = celltypes)) +
  scale_fill_manual(values = col) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  theme(legend.position = "none")

```


```#{r, fig.width = 10}
peakset %>%# group_by(GroupReplicate) +
  ggplot() +
  geom_bar(aes(x = GroupReplicate, fill = GroupReplicate)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  theme(legend.position = "none")
```

The number of peaks for each celltype

```{r, fig.width=10, fig.height=10}
p1 <- peakset %>% 
  separate(col = GroupReplicate, into = c("celltypes","replicate"), sep = "\\._\\.", 
           remove = FALSE) %>%
  ggplot() +
  geom_bar(aes(x = celltypes, fill = celltypes)) +
  scale_fill_manual(values = col) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  theme(legend.position = "none") +
  labs(title = "Number of peaks for each celltype")

p2 <- getCellColData(proj) %>% 
  as.data.frame()%>% 
  ggplot() +
  geom_bar(aes(x = celltypes, fill = celltypes)) +
  scale_fill_manual(values = col) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  theme(legend.position = "none") +
  labs(title = "Number of cells in each celltpye")

gridExtra::grid.arrange(p1, p2, ncol = 1)
```

```{r}
test %>% group_by(celltypes) %>% summarize(n = n())

```

