---
title: "Peak2Gene links and gene activity score"
output: 
  html_document:
    toc: true
    toc_depth: 5
    code_folding: hide
    toc_float: true
    code_download: true
    theme: cosmo
    highlight: textmate
---

<style>
body {
text-align: justify}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = FALSE, autodep = TRUE, 
                      collapse = TRUE, message = FALSE)
knitr::opts_knit$set(root.dir = "/omics/groups/OE0533/internal/katharina/scDoRI/gastrulation_data/")
setwd("/omics/groups/OE0533/internal/katharina/scDoRI/gastrulation_data/")

set.seed(1)
```

  
```{r}
## Load libraries


suppressPackageStartupMessages({
library(ArchR)
library(rhdf5)
library(tidyverse)
library(reticulate)
library(zellkonverter)
library(Matrix)
library(dichromat)
#library(caret)
h5disableFileLocking()})
```

# Create peak-to_gene link matrix

ArchR returns a dataframe containg all peak-to-gene links and metadata with
GRanges objects. We want to create a peak-to-gene matrix with chromosome & peak
index as column names and gene names as rownames.



The p2g dataframe has a metadata component which contains a Granges object of 
the relevant peaks. The idxATAC column of the p2g dataframe corresponds to this 
Granges object. Below we extract the chr & chr_idx for later. We will also 
extract the gene names based on the idx. 

```{r}
#metadata_p2g_atac <- as.data.frame(metadata(p2g_original)[[1]])
#metadata_p2g_rna <- as.data.frame(metadata(p2g_original)[[2]])


p2g_matrix_from_df <- function(p2g, metadata_p2g_rna, metadata_p2g_atac){
  # add chrosome identifiers to p2g dataframe
  metadata_p2g_atac <- metadata_p2g_atac %>%  rownames_to_column("idxATAC") %>% 
  group_by(seqnames) %>% 
  # create idx per chromosome
  mutate(idx = seq_along(seqnames)) %>% ungroup() %>% 
  # create chr & index combination column, will be needed, since this is how we name peak matrix
  tidyr::unite(chr_idx, seqnames, idx, remove = FALSE, sep = "_")
  
  
  #fitler to get only the peaks which we have in the p2g links dataframe
  atac_chr_idx <- metadata_p2g_atac %>% filter(idxATAC %in% p2g$idxATAC) %>%
    select(idxATAC, chr_idx) 
  
  atac_chr_idx$idxATAC <- as.integer(atac_chr_idx$idxATAC)
  
  # now we can merge this with the p2g dataframe based on the idxATAC column
  p2g_idx <- inner_join(atac_chr_idx, p2g, by = "idxATAC",copy = TRUE)
  
  # add gene names to p2g dataframe
  # get gene names
  rna_idx <- metadata_p2g_rna %>% rownames_to_column("idxRNA") %>% 
    group_by(seqnames) %>% 
    mutate(idx = seq_along(seqnames)) %>% ungroup() %>% 
    tidyr::unite(chr_idx_rna, seqnames, idx, remove = FALSE, sep = "_") %>% 
    filter(idxRNA %in% p2g_idx$idxRNA) %>% 
    select(idxRNA, name, chr_idx_rna) %>% rename("gene" = name)
  
  rna_idx$idxRNA <- as.integer(rna_idx$idxRNA)
  
  # combine this with the p2g dataframe
  p2g_idx <- inner_join(rna_idx, p2g_idx, by = "idxRNA", copy = TRUE)
  
  # create p2g link matrix
  # create column and row indices for the unique peak and rna indices
  p2g_idx$ColIndex <- match(p2g_idx$chr_idx, unique(p2g_idx$chr_idx))
  p2g_idx$RowIndex <- match(p2g_idx$gene, unique(p2g_idx$gene))
  
  # create a sparse matrix with the correlation values
  p2g_mat <- sparseMatrix(i = p2g_idx$RowIndex, 
                          j = p2g_idx$ColIndex,
                          x = p2g_idx$Correlation,
                          dims = c(length(unique(p2g_idx$gene)),
                                   length(unique(p2g_idx$chr_idx))),
                          dimnames = list(unique(p2g_idx$gene), 
                                          unique(p2g_idx$chr_idx)))
  
  print(paste0("The sparse matrix of correlations has dimensions: ", dim(p2g_mat)[[1]], " genes and ", dim(p2g_mat)[[2]], " peaks."))
  print(paste0("Minimum value: ", min(p2g_mat)))
  print(paste0("The maximum value: ", max(p2g_mat)))
  
  return(list(p2g_mat, p2g_idx))
}
```





# Gene activity scores 

```{r}
gene_activity_scores <- function(peak_mat, p2g_mat) {
  peak_mat_subset <- peak_mat[colnames(p2g_mat), ]
  # normalize the p2g matrix by the total number of peaks linked to each gene
  p2g_mat <- p2g_mat / rowSums(p2g_mat)
  print(paste0("normalized the p2g matrix"))
  # Now we can compute a weighted sum of peak2gene correlations for each
  # peak and gene
  scores <- p2g_mat %*% peak_mat_subset
  print(paste0("Computed weightes sum of peaks for each gene and cell"))
  # create a dataframe for computing the linear model
  linear_model_df <- data.frame(cell = colnames(scores),
                               total_activity = colSums(scores),
                               total_sites = colSums(peak_mat_subset))
  # compute a linear model
  activity_model <- stats::lm(log(total_activity) ~ log(total_sites),
                            data = linear_model_df)
  # extract the fitted model
  linear_model_df$fitted_curve <- exp(as.vector(predict(activity_model,
                                                         type = "response")))
  # compute size factors from fitted model
  size_factors <- mean(linear_model_df$fitted_curve) / linear_model_df$fitted_curve
  # create diagonal matrix containing the size factors
  size_factors_mat <- Matrix::Diagonal(x = size_factors)
  row.names(size_factors_mat) <- linear_model_df$cell
  # normalize by library depth size factors
  norm_scores <- Matrix::t(size_factors_mat %*% Matrix::t(scores))
  print(paste0("Normalized for library size"))
  # exponentiate, because RNA counts are log-normally distributed
  norm_scores@x <- pmin(1e9, exp(norm_scores@x) - 1)
  print(paste0("Exponentiated matrix"))
  
  # free some memory
  rm(peak_mat_subset)
  rm(activity_model)
  rm(scores)
  gc(reset = TRUE)

  # scale with total activity scores again
  scale_factors <- Matrix::Diagonal(x = 1/Matrix::colSums(norm_scores))
  print(paste0("Divided by total activity to get value between zero and one"))
  
  final_scores <- Matrix::t(scale_factors %*% Matrix::t(norm_scores))

  return(final_scores)

}

```




# Gene activity score computation

Now we will calculate a weighted sum of accessible peaks in a cell. The weights
are the corresponding correlation values between a particular peak $j$ and gene
$k$. $A\in\mathbb{R}^{JxI}$ is the binary peak accessibility matrix with $A_{ji} = 1$
if peak $j$ is found in cell $i$ and  $A_{ji} = 0$ otherwise.

$R_{ki} = \sum{U_{kj}A_{ji}}$, where $R_{ki}$ corresponds to the gene activity 
score of gene $k$ in cell $i$.

$R = U I$ with $R \in \mathbb{R}^{KxI}$. 

Below you can see the first few entries of this gene activity score matrix. Note 
that some entries are negative.

Try to simply divide by row sums: $\frac{U_{kj}}{\sum_{j}{U_{kj}}}$.



### Linear regression to account for differences in library size.

The aggregate correlation values will depend on scATAC-seq library depth in each 
cell. We can capture this relationship with a linear regression. 

$\log{\sum_{k}{r_k} = \beta_0 + \beta_A \log(\sum_j{A_{ji}})}$


### SizeFactors

The output of the linear model is used to scale the accessibility we computed
above.

$\tilde{R}_{ki}= R_{ki} * \frac{\sum_{i}{r_i}}{r_i}$

If an activity score $R_{ki}$ for a given cell $i$ is smaller than the average activity score, the 
size factor will be >1 and if the activity score  is bigger than the average 
activity score the size factor will be <1. After correcting with the size factors we would expect the linear increase in activity with increasing number of accessibile 
sites to be corrected for. 


### Transform

Then we exponentiate the values, because the RNA-seq measures are typically 
log-normally distributed. To create relative exponentiated gene accessibility
values, we scale them by the total exponentiated gene accesibility values. We 
get our gene activity scores $G_{ki}$.

Because for log transformation of gene expression counts we add a pseudocount, 
we need to subtract one here:

$e^{\tilde{R}} - 1$

$G_{ki} = \frac{e^{\tilde{R}_{ki}}-1}{\sum_k{e^{\tilde{R}_{ki}}-1}}$



### Scaling by total gene activity in a given cell

Here, you can see the first few entries of the transformed matrix. The column
sum is now 1. 










